### ISSUE TITLE
How can Chrome's memory be reduced!!!!

### ISSUE THREAD

**Issue by donghang1994** [emscripten-core/emscripten#15591]
![image](https://user-images.githubusercontent.com/91657327/142830722-6782e22b-a6db-4214-b6e0-8c26f0455456.png)

C++ Memory has been freed, but Browser memory usagedoes not come down.
Don't you have this problem..

**Comment by sbc100** [comment]
Are you using growable memory?  This is the only way that you would see C/C++ memory allocations increase the usage here.. otherwise memory is all allocated up front when the program starts and never changes.

If you are using growable memory, sadly there is no way to ever shrink it.

**Comment by kripken** [comment]
If that's not it, the devtools heap profiler might find something on the JS side. Could be something leaking there.

**Comment by donghang1994** [comment]
> Are you using growable memory? This is the only way that you would see C/C++ memory allocations increase the usage here.. otherwise memory is all allocated up front when the program starts and never changes.
> 
> If you are using growable memory, sadly there is no way to ever shrink it.

I use "ALLOW_MEMORY_GROWTH=1", you mean the browser  memory can't be lowered even if my c++ has freed it.

**Comment by sbc100** [comment]
Yes, WebAssembly memories can (sometimes) be grown, but as of today there is no way to ever shrink them.   If you would like some more background on this there is fairly long discussion here: https://github.com/WebAssembly/design/issues/1397

Of course, within the WebAssembly memory you still do need to use memory management such as new and delete to avoid memory leaks within the memory.

**Comment by donghang1994** [comment]
> If that's not it, the devtools heap profiler might find something on the JS side. Could be something leaking there.

How do I use it?

**Comment by donghang1994** [comment]
> Yes, WebAssembly memories can (sometimes) be grown, but as of today there is no way to ever shrink them. If you would like some more background on this there is fairly long discussion here: [WebAssembly/design#1397](https://github.com/WebAssembly/design/issues/1397)
> 
> Of course, within the WebAssembly memory you still do need to use memory management such as new/delete/malloc/free to avoid memory leaks within the memory.

But then my browser's memory will keep growing,Won't it make andy difference,

**Comment by donghang1994** [comment]
> Yes, WebAssembly memories can (sometimes) be grown, but as of today there is no way to ever shrink them. If you would like some more background on this there is fairly long discussion here: [WebAssembly/design#1397](https://github.com/WebAssembly/design/issues/1397)
> 
> Of course, within the WebAssembly memory you still do need to use memory management such as new/delete/malloc/free to avoid memory leaks within the memory.

But then my browser's memory will keep growing,Won't it make any difference,

**Comment by donghang1994** [comment]
> Yes, WebAssembly memories can (sometimes) be grown, but as of today there is no way to ever shrink them. If you would like some more background on this there is fairly long discussion here: [WebAssembly/design#1397](https://github.com/WebAssembly/design/issues/1397)
> 
> Of course, within the WebAssembly memory you still do need to use memory management such as new/delete/malloc/free to avoid memory leaks within the memory.

![BaiduShurufa_2021-11-23_10-43-13](https://user-images.githubusercontent.com/91657327/142963937-5a82ba27-9d3e-4c70-9b24-31b51a6f25d3.png)
Is this error due to memory?

**Comment by sbc100** [comment]
Does your program actually use more than 2Gb of heap?   That fact that its going up that high suggests it is most likely not free'ing the existing memory within the heap.

**Comment by sbc100** [comment]
Do you need `ALLOW_MEMORY_GROWTH`?   Many types of programs should not need that flag since they can run with a pre-defined heap size and take case of free'ing (and re-using) the memory within the heap.

**Comment by donghang1994** [comment]
> 

If I don't use ALLOW_MEMORY_GROWTH, Is there any problem with actually using more than the defined heap size.
And use ALLOW_MEMORY_GROWTH, why not enough memory,  report "Cannot enlarge memory".

**Comment by sbc100** [comment]
If you don't define `ALLOW_MEMORY_GROWTH` then your heap will be of a fixed size.  If your program is will behaved this is often fine, you just need to manage your allocations and unless you are doing some operation that requires ever increasing memory your program should run just find within its bounds. 

Are you performing operations that you require large or unbounded amounts of memory?  Or do you have memory leaks?   

If you don't are no dealing with huge allocations, then you probably have memory leaks and you should not use `ALLOW_MEMORY_GROWTH` and instead set your `TOTOL_MEMORY` to something small so that you hit the limit quickly (that should help you find your memory leaks).  You might also want to try `-fsanitize=leak` which can help you find leaks.

Do you have a native version of your application?  Does it leak memory or use very large amounts?

**Comment by donghang1994** [comment]
> 

It use very large amounts, No memory leaks. So there's nothing can do about it?

**Comment by sbc100** [comment]
So you use a very large amount of memory temporarily and then you want to be able to reduce the browser memory usage once you are done using it?    No, there is no way to do that today in WebAssembly without discarding the instance and creating a new one.  Is that an option for you?    The memory is tied to the instance and can never shrink, only grow.

**Comment by donghang1994** [comment]
What do you mean of 'discarding  instance '. If I switch pages in my browser app,is there anything I can do to free up memory before switching.  Because I don't use anything from the last page anymore.

**Comment by sbc100** [comment]
If you use `-sMODULARIZE` you can create (and discard) instances of your application more easily.   So you would do `var mymodule = Module()` to create a new one.   `mymodule = null` would then discard the reference to it and should allow chrome to free all the memory associated with it. 

See https://emscripten.org/docs/getting_started/FAQ.html#can-i-use-multiple-emscripten-compiled-programs-on-one-web-page

**Comment by curiousdannii** [comment]
I think there are some problems currently where the event handlers mean that modules won't be correctly garbage collected.

**Comment by donghang1994** [comment]
> I think there are some problems currently where the event handlers mean that modules won't be correctly garbage collected. See #9820

Wasm memory will increase as it is used， It has been there and it hasn't been solved
